synapse/handlers/message.py

```diff
@@ -225,6 +239,89 @@ def get_joined_members(self, requester, room_id):
+            yield self._schedule_next_expiry()
+
+        next_scheduled_expiry_ts = self._scheduled_expiry.getTime() * 1000
+        if expiry_ts < next_scheduled_expiry_ts:
+            self._scheduled_expiry.cancel()
+            yield self._schedule_next_expiry()
+
+    @defer.inlineCallbacks
+    def _schedule_next_expiry(self):
+        """Retrieve the expiry timestamp of the next event to be expired, and schedule
+        an expiry task for it.
+
+        If there's no event left to expire, set _expiry_scheduled to False so that a
+        future call to save_expiry_ts can schedule a new expiry task.
+        """
+        # Try to get the expiry timestamp of the next event to expire.
+        next_event_to_expire = yield self.store.get_next_event_to_expire()
+
+        if next_event_to_expire:
+            # Figure out how many seconds we need to wait before expiring the event.
+            now_ms = self.clock.time_msec()
+            delay = (next_event_to_expire["expiry_ts"] - now_ms) / 1000
+
+            # callLater doesn't support negative delays, so trim the delay to 0 if we're
+            # in that case.
+            if delay < 0:
+                delay = 0
+
+            logger.info(
+                "Scheduling expiry of event %s in %.3fs",
+                next_event_to_expire["event_id"],
+                delay,
+            )
+
+            self._scheduled_expiry = self.clock.call_later(delay, self._expire_event)
+        else:
+            # If there's no more event to expire, then set _expiry_scheduled to False, so
+            # that the next call to save_expiry_ts can schedule a new expiry task.
+            self._scheduled_expiry = False
+
+    @defer.inlineCallbacks
+    def _expire_event(self):
+        """Retrieve and expires an event that needs to be expired from the database.
+
+        If the event doesn't exist in the database, log it and delete the expiry date
+        from the database (so that we don't try to expire it again).
+        """
+        if not self._ephemeral_events_enabled:
+            return
+
+        # Get the ID of the next event to expire.
+        next_event_to_expire = yield self.store.get_next_event_to_expire()
+        event_id = next_event_to_expire["event_id"]
+
+        logger.info("Expiring event %s", event_id)
+
+        # Try to retrieve the event from the database.
+        event = yield self.store.get_event(event_id)
+
+        if not event:
+            # If we can't find the event, log a warning and delete the expiry date from
+            # the database so that we don't try to expire it again in the future.
+            logger.warning("Can't expire event %s because we don't have it." % event_id)
+            yield self.store.delete_event_expiry(event)
+            return
+
+        # Expire the event. This function also deletes the expiry date from the database
+        # in the same database transaction.
+        yield self.store.expire_event(event)
```

Comment:
I would move all of this into `store.expire_event`; I think you could avoid the need to expose `delete_event_expiry` as a separate function.